src/sympc/module/nn/functional.py

```diff
@@ -37,3 +49,327 @@ def mse_loss(pred: MPCTensor, target: MPCTensor, reduction: str = "mean") -> MPC
+
+    tensor_type = session.tensor_type
+    tensor = torch.zeros(res_shape, dtype=tensor_type)
+
+    for i in range((res_shape[-2] - kernel_size[0]) // stride[0] + 1):
+        row_idx = i * stride[0]
+        for j in range((res_shape[-1] - kernel_size[1]) // stride[1] + 1):
+            col_idx = j * stride[1]
+            if len(res_shape) == 4:
+                tensor[
+                    :,
+                    :,
+                    row_idx : row_idx + kernel_size[0],
+                    col_idx : col_idx + kernel_size[1],
+                ] += grads_share.tensor[:, :, i, j]
+            else:
+                tensor[
+                    :,
+                    row_idx : row_idx + kernel_size[0],
+                    col_idx : col_idx + kernel_size[1],
+                ] += grads_share.tensor[:, i, j]
+
+    if len(res_shape) == 4:
+        tensor = tensor[
+            :, :, padding[0] : input_shape[-2], padding[1] : input_shape[-1]
+        ]
+    else:
+        tensor = tensor[
+            :,
+            padding[0] : res_shape[-2] - padding[0],
+            padding[1] : res_shape[-1] - padding[1],
+        ]
+    res = ShareTensor(config=grads_share.config)
+    res.tensor = tensor
+
+    return res
+
+
+def max_pool2d_backward(
+    grad: MPCTensor,
+    input_shape: Tuple[int],
+    indices: MPCTensor,
+    kernel_size: Union[int, Tuple[int, int]],
+    stride: Optional[Union[int, Tuple[int, int]]] = None,
+    padding: Union[int, Tuple[int, int]] = 0,
+    dilation: Union[int, Tuple[int, int]] = 1,
+) -> MPCTensor:
+    """Helper function for the backwards step for max_pool2d.
+
+    Credits goes to the CrypTen team.
+
+    Args:
+        grad (MPCTensor): gradient that comes from the child node
+        input_shape (Tuple[int]): the shape of the input when the max_pool2d was run
+        indices (MPCTensor): the indices where the maximum value was found in the input
+        kernel_size (Union[int, Tuple[int, int]]): the kernel size
+            in case it is passed as an integer then that specific value is used for height and width
+        stride (Union[int, Tuple[int, int]]): the stride size
+            in case it is passed as an integer then that specific value is used for height and width
+        padding (Union[int, Tuple[int, int]]): the padding size
+            in case it is passed as an integer then that specific value is used for height and width
+        dilation (Union[int, Tuple[int, int]]): the dilation size
+            in case it is passed as an integer then that specific value is used for height and width
+
+    Returns:
+        The gradient that should be backpropagated (MPCTensor)
+
+    Raises:
+        ValueError: In case some of the values for the parameters are not supported
+    """
+    kernel_size, stride, padding, dilation = _sanity_check_max_pool2d(
+        kernel_size, stride, padding, dilation
+    )
+    if len(grad.shape) != 4 and len(grad.shape) != 3:
```

Comment:
Shouldn't it be or? instead of and?

Reply:
What we want to check here if `grad` shape is 3 or 4 -- the if will be triggered only if it is neither 3 or 4.
If we apply De Morgan rule (with double negation) we could have an or - but I think it is ok to have it with an `and`.

Or do you think it is not understandable what was the intention here?